iotfunctions/pipeline.py

```diff
@@ -2304,14 +2305,14 @@ def _execute_stage(self, stage, df, start_ts, end_ts, entities, register, to_csv
 
         start_time = pd.Timestamp.utcnow()
         try:
-
+            scope_mask = self.apply_scope(df, stage)
             contains_extended_args = self._contains_extended_arguments(stage.execute)
-
             if contains_extended_args:
-                newdf = stage.execute(df=df, start_ts=start_ts, end_ts=end_ts, entities=entities)
+                newdf = stage.execute(df=df[scope_mask], start_ts=start_ts, end_ts=end_ts, entities=entities)
             else:
-                newdf = stage.execute(df=df)
-
+                newdf = stage.execute(df=df[scope_mask])
+            cols_to_merge = newdf.columns.difference(df.columns)
+            newdf = df.merge(newdf[cols_to_merge], how='left', left_index=True, right_index=True)
```

Comment:
@pkohlmann This is based on the assumption that every time the pipeline runs a stage, it will always create output columns.

1. Let’s say original_df (in this case df)  has column `x,y,z`  with `n rows`.

1. And the modified_df (in this case newdf) has columns` x,y,z,a,b` with less than `n rows`(because of scope filter).  `a,b` are new columns created for derived metrics

1. Final dataframe will have columns  `x,y,z,a,b` with `n rows`

Is there a chance that x,y,z  columns from modified_df also change ?

Reply:
there are many issues can happen here. 
1) if user messes up the dataframe index then the merge is not possible. 
2) What happens in case of aggregation? The merge on index will have issues. Because the index value may be different. 
3) as per your 3rd point there is a chance that you will have only n rows. That is not really true. if someone calls loaders then you may have an n+ rows. 
4) Also please move the merge feature to a different function. It will be a complicated function. Many things go wrong.  Have proper try and catch mechanism.

Reply:
@herleraja 
Noted 
I'll incorporate changes in the next PR.
Aggregation as we have decided will be dealt later